src/coprocessor/dag/expr/builtin_string.rs

```diff
@@ -683,32 +681,131 @@ impl ScalarFunc {
+            }
         }
-        // check max_allowed_packet when it's pushed down, add warning if needed
-        if !target_len_positive
-            || target_len > cop_datatype::MAX_BLOB_WIDTH as usize
-            || (pad.is_empty() && input.len() < target_len)
-        {
-            return Ok(None);
+    }
+
+    #[inline]
+    pub fn lpad<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let input = try_opt!(self.children[0].eval_string_and_decode(ctx, row));
+        let target_len = try_opt!(self.children[1].eval_int(ctx, row));
+        let pad = try_opt!(self.children[2].eval_string_and_decode(ctx, row));
+        let input_len = input.chars().count();
+
+        match validate_target_len_for_pad(
+            self.children[1].field_type(),
+            target_len,
+            input_len,
+            4,
+            pad.is_empty(),
+        ) {
+            None => Ok(None),
+            Some(0) => empty_str(),
+            Some(target_len) => {
+                let r = if let Some(remain) = target_len.checked_sub(input_len) {
+                    pad.chars()
+                        .cycle()
+                        .take(remain)
+                        .chain(input.chars())
+                        .collect::<String>()
+                } else {
+                    input.chars().take(target_len).collect::<String>()
+                };
+                Ok(Some(Cow::Owned(r.into_bytes())))
+            }
         }
+    }
 
-        let r = input
-            .iter()
-            .chain(pad.iter().cycle())
-            .cloned()
-            .take(target_len)
-            .collect::<Vec<_>>();
-        Ok(Some(Cow::Owned(r)))
+    #[inline]
+    pub fn lpad_binary<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let input = try_opt!(self.children[0].eval_string(ctx, row));
+        let target_len = try_opt!(self.children[1].eval_int(ctx, row));
+        let pad = try_opt!(self.children[2].eval_string(ctx, row));
+
+        match validate_target_len_for_pad(
+            self.children[1].field_type(),
+            target_len,
+            input.len(),
+            1,
+            pad.is_empty(),
+        ) {
+            None => Ok(None),
+            Some(0) => empty_str(),
+            Some(target_len) => {
+                let r = if let Some(remain) = target_len.checked_sub(input.len()) {
+                    pad.iter()
+                        .cycle()
+                        .take(remain)
+                        .chain(input.iter())
+                        .cloned()
+                        .collect::<Vec<_>>()
+                } else {
+                    input[..target_len].to_vec()
+                };
+                Ok(Some(Cow::Owned(r)))
+            }
+        }
     }
 }
 
+fn empty_str() -> Result<Option<Cow<'static, [u8]>>> {
```

Comment:
I personally prefer not to introducing a function for this simple logic, although I'm fine with it.